### PR TITLE
RFC: allow reexecuting generator step and any of its downstream together

### PR DESCRIPTION
**Pull request by yuhan**
### Summary & Motivation

This proposal is going to result in a user-facing behavioral change - it allows additional possible cases (see test case `test_select_dynamic_step_with_non_static_mapping`) and doesn't affect existing patterns, which could resolve https://github.com/dagster-io/dagster/issues/6044

The fix here is to allow re-executing the step that generates dynamic outs (aka `generate_subtasks`) and any downstream dynamic steps that were resolved from those dynamic outs (e.g. `subtask[4]`), which would impose the following logics:
image
- When `generate_subtasks` aren't always generating the same dynamic outs in each re-execution, the execution engine would use the execution plan that's constructed in **the same run** but only execute the selected steps. It means if a selected step is not being constructed in `generate_subtasks`, it will skip because its upstream output doesn't exist.

Alternative is ... TODO

### How I Tested These Changes
- Repro locally: all cases mentioned on the issue work now
```
# didn't work and now work
+hello_dynamic
generate_subtasks*
generate_subtasks+
generate_subtasks, hello_dynamic
# did work and still work
hello_dynamic
hello_dynamic*
hello_dynamic[subtask_1]
```
image

- Added more unit tests for edge cases

### CODE REVIEW

File: python_modules/dagster/dagster/core/execution/plan/plan.py [hunk 1]
```diff
@@ -758,19 +758,32 @@ def build_subset_plan(
...
+
+        for key in step_keys_to_execute:
+            handle = StepHandle.parse_from_key(key)
+
             if handle not in self.step_dict:
+                if isinstance(handle, ResolvedFromDynamicStepHandle):
+                    unresolved_handle = UnresolvedStepHandle(solid_handle=handle.solid_handle)
+                    if unresolved_handle in self.step_dict: # ok if the unresolved version is present
+                        step_handles_to_execute_set.add(unresolved_handle)
+                        continue
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/execution/plan/plan.py:772]: add some comments to clarify the extrapolation happening here

**Review by alangenfeld — CHANGES_REQUESTED** [review]
i am cool with this once it includes tests

File: python_modules/dagster/dagster/core/execution/plan/plan.py (hunk 1 above)
Comment (yuhan) [reply python_modules/dagster/dagster/core/execution/plan/plan.py]: After further investigation, i found the root cause isn't about the parsing or unresolved vs resolved. the root cause is that `self.step_dict` here doesn't have a bunch of `ResolvedFromDynamicStepHandle` that is selected here - and this is because in [1] we explicitly dropped the steps we intent to re-execute. So this may have not been a bug but a feature and I think we should loosen that constraints (i.e. not drop these steps when generating dynamic_mappings from parent_state) and allow [2]

File: python_modules/dagster/dagster/core/execution/plan/state.py [hunk 2]
```diff
@@ -120,4 +120,5 @@ def for_reexecution(
             for step_key in parent_state.dynamic_mappings.keys()
             if step_key not in step_keys_to_execute
         }
-        return KnownExecutionState({}, dynamic_mappings_to_use)
+        # return KnownExecutionState({}, dynamic_mappings_to_use)
+        return KnownExecutionState({}, parent_state.dynamic_mappings)
```
Comment (yuhan) [inline python_modules/dagster/dagster/core/execution/plan/state.py:124]: [1]

File: python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_dynamic_execution.py [hunk 3]
```diff
@@ -283,6 +283,66 @@ def test_fan_out_in_out_in(run_config):
...
+            instance=instance,
+            step_selection=["emit", "emit_ten", "multiply_inputs[1]", "multiply_by_two[1]"],
+        )
+
+        keys_5 = result_5.events_by_step_key.keys()
+        assert "multiply_inputs[1]" in keys_5
+        assert "multiply_by_two[1]" in keys_5
+        assert "multiply_inputs[0]" not in keys_5
+
 def test_bad_step_selection():
```
Comment (yuhan) [inline python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_dynamic_execution.py:346]: [2]
```
step_selection=["emit", "multiply_by_two[1]"]
```

File: python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_dynamic_execution.py [hunk 4]
```diff
@@ -283,20 +287,140 @@ def test_fan_out_in_out_in(run_config):
...
 
-        # this exact error could be improved, but it should fail if you try to select
-        # both the dynamic outputting step key and something resolved by it in the previous run
-        with pytest.raises(DagsterExecutionStepNotFoundError):
-            reexecute_pipeline(
-                dynamic_pipeline,
-                parent_run_id=result_1.run_id,
-                instance=instance,
-                step_selection=["emit", "multiply_by_two[1]"],
-            )
```
Comment (yuhan) [inline python_modules/dagster/dagster_tests/execution_tests/dynamic_tests/test_dynamic_execution.py:299]: the new pattern should be allowing this

**Comment by yuhan** [comment]
close in favor of https://github.com/dagster-io/dagster/pull/7979
